pdns/recursordist/pdns_recursor.cc

```diff
@@ -1846,7 +1846,7 @@ void startDoResolve(void* arg) // NOLINT(readability-function-cognitive-complexi
     }
 
     if (comboWriter->d_mdp.d_header.opcode == static_cast<unsigned>(Opcode::Query)) {
-      if (resolver.d_outqueries != 0 || resolver.d_authzonequeries != 0) {
+      if (res == RCode::ServFail || resolver.d_outqueries != 0 || resolver.d_authzonequeries != 0) {
```

Comment:
While I agree with that the current behaviour is not right, we cannot simply rely on the rcode. We do serve `ServFail` answers from the cache, for example when a DNSSEC proof is broken, so a `ServFail` is not always a cache miss.

Reply:
Indeed, will look into a better condition. 

Also a nte to @zjs604381586: we prefer PRs that do not contain merge commits Start with a clean, up-to-date master, branch and only add commits relevant to the fix. Then use the branch to submit the PR.

Reply:
> Indeed, will look into a better condition.
> 
> Also a nte to @zjs604381586: we prefer PRs that do not contain merge commits Start with a clean, up-to-date master, branch and only add commits relevant to the fix. Then use the branch to submit the PR.

I see, thank you